hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/ReadableKVState.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.spi.state;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.Iterator;
+
+/**
+ * Provides access to key/value state for a service implementation. This interface is implemented by
+ * the Hedera application, and provided to the service implementation at the appropriate times. The
+ * methods of this class provide read access to the state.
+ *
+ * <p>Null values <strong>cannot be stored</strong> in this state. Null is used to indicate the
+ * absence of a value.
+ *
+ * @param <K> The type of the key
+ * @param <V> The type of the value
+ */
+public interface ReadableKVState<K extends Comparable<K>, V> {
```

Comment:
Wow! It looks like that actually works. This would be a very strong guarantee that the returned objects are immutable. This is great!

Reply:
At least in theory, our records could reference mutable objects. But it would still be a strong indication to developers, that `V` is expected to be immutable.

Reply:
I really like this idea, but let's wait until we get the PBJ objects integrated so we can then make the switch without having a huge changeset across existing code.

Reply:
https://github.com/hashgraph/hedera-services/issues/4614